### PR TITLE
Fix mount option issue in  rbd

### PR DESCRIPTION
**Pull request by Madhu-1**
use mountoptions when mounting rbd to stagingpath in stagevolume request, add E2E for mount options

fixes: #846
updates: #757

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
logs
```text
sh-4.2# cat /proc/mounts |grep rbd |grep pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9
/dev/rbd1 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9/globalmount/0001-0004-abcd-0000000000000003-576d87b3-5c5c-11ea-964b-0a7db111c845 ext4 rw,seclabel,relatime,discard,stripe=1024,data=ordered 0 0
/dev/rbd1 /var/lib/kubelet/pods/30451fe1-64a7-4456-9b32-02811ae79348/volumes/kubernetes.io~csi/pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9/mount ext4 rw,seclabel,relatime,discard,stripe=1024,data=ordered 0 0
sh-4.2# 
sh-4.2# 
sh-4.2# cat /proc/mounts |grep pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9
/dev/rbd1 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9/globalmount/0001-0004-abcd-0000000000000003-576d87b3-5c5c-11ea-964b-0a7db111c845 ext4 rw,seclabel,relatime,discard,stripe=1024,data=ordered 0 0
/dev/rbd1 /var/lib/kubelet/pods/30451fe1-64a7-4456-9b32-02811ae79348/volumes/kubernetes.io~csi/pvc-1185e1a3-35a1-4b2d-b559-b83d3b3686c9/mount ext4 rw,seclabel,relatime,discard,stripe=1024,data=ordered 0 0
[🎩︎]mrajanna@localhost k8s-cluster $]kubectl exec -it csi-rbd-demo-pod1 sh
# mount |grep /var/lib/www/html
/dev/rbd1 on /var/lib/www/html type ext4 (rw,relatime,seclabel,discard,stripe=1024,data=ordered)
# 

```

**Comment by Madhu-1** [comment]
storage usage
* before
```
dd if=/dev/zero of=file.txt count=1024 bs=1048576
dd: error writing 'file.txt': No space left on device
958+0 records in
957+0 records out
1003896832 bytes (1.0 GB, 957 MiB) copied, 4.97333 s, 202 MB/s
# df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/rbd0       976M  960M     0 100% /var/lib/www/html
```
rbd image disk usage
```text
sh-4.2# rbd du csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845 --pool=replicapool
warning: fast-diff map is not enabled for csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845. operation may be slow.
NAME                                         PROVISIONED USED    
csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845       1 GiB 988 MiB 
```
* After
```text
# ls
file.txt  lost+found
# rm -rf file.txt
# df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/rbd0       976M  2.6M  958M   1% /var/lib/www/html
```
rbd image disk usage
```text
rbd du csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845 --pool=replicapool
warning: fast-diff map is not enabled for csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845. operation may be slow.
NAME                                         PROVISIONED USED   
csi-vol-576d87b3-5c5c-11ea-964b-0a7db111c845       1 GiB 40 MiB 
```

**Comment by mykaul** [comment]
I somehow remember we discussed this cannot work because of some privileges issue - @dillaman  might remember.

**Comment by dillaman** [comment]
> I somehow remember we discussed this cannot work because of some privileges issue - @dillaman might remember.

I think that was in reference to running `fstrim` within the pod -- which is disallowed. It's still potentially not a good idea to run in online discard mode for RBD since the ext4/XFS file systems are not very efficient w/ it since they expect fast, local storage that basically doesn't block the IO path.

**Comment by Madhu-1** [comment]
> > I somehow remember we discussed this cannot work because of some privileges issue - @dillaman might remember.
> 
> I think that was in reference to running `fstrim` within the pod -- which is disallowed. It's still potentially not a good idea to run in online discard mode for RBD since the ext4/XFS file systems are not very efficient w/ it since they expect fast, local storage that basically doesn't block the IO path.

@dillaman is there any better way to solve this issue?

**Comment by humblec** [comment]
@Madhu-1 the `mountOptions` construction happens on staging path here, we do that for publish path too. iic, in my testing I have seen CSI log says, the mount options whatever we supplied has gone to "publish path", why its not getting applied is a question for me . The other interesting part here is that, why do we want to do "mount options" for both "staging" and "publishing" ?, I read some where that, we have to do it for both, but then the question would be , Is it a `bind mount` limitation or some bug in `CSI` itself. @gnufied any thoughts here ?

**Comment by humblec** [comment]
Thinking some more on the bind mounts, cant we do `shared` mount propagation to solve the issue instead of doing it separately on `staging and publishing ` ? this is based on my assumption that, shared mount propagation allows us to work both paths source and target with exactly same mount options. At the same time I am not sure the `shared` mount propagation can cause some other issues considering staging is our global mount path and we have separate target paths at time of publish, and  the publish may require different view of the mount path at publish time.   I am trying to understand what cause the missing options and why do we have to do it for both and possibly why that requirement arise?

**Comment by humblec** [comment]
@dillaman any thoughts on bind mount or the points raised in above comments ?

**Comment by humblec** [comment]
Approving this PR as the fix is in effect. If there are some improvement we can do, it can be in follow up PR too.

**Comment by humblec** [comment]
@nixpanic any thoughts from your side on the bind mount ?

**Comment by dillaman** [comment]
> @nixpanic any thoughts from your side on the bind mount ?

I don't think it matters for the bind mount (if that is your concern). You would want the node stage mount point to have the options since that is where the file system is initialized and controlled. The node publish bind mount is then just essentially reference counting.

File: pkg/rbd/nodeserver.go [hunk 1]
```diff
@@ -392,6 +392,7 @@ func (ns *NodeServer) mountVolumeToStagePath(ctx context.Context, req *csi.NodeS
 	}
 
 	opt := []string{"_netdev"}
+	opt = csicommon.ConstructMountOptions(opt, req.GetVolumeCapability())
```
Comment (nixpanic) [inline pkg/rbd/nodeserver.go]: you are overriding the previous `"_netdev"` option, that is probably not intentional.

File: e2e/utils.go [hunk 2]
```diff
@@ -992,3 +992,39 @@ func pvcDeleteWhenPoolNotFound(pvcPath string, cephfs bool, f *framework.Framewo
...
+	err = createPVCAndApp("", f, pvc, app)
+	if err != nil {
+		return err
+	}
+
+	opt := metav1.ListOptions{
+		LabelSelector: "app=validate-mount-opt",
+	}
+
+	mountOpt := "discard"
```
Comment (nixpanic) [inline e2e/utils.go]: I do not like the fact that the `discard` option is hardcoded here. Can it be checked that the option is also part of the PVC request? Or pass the mount option as an argument from the test?

File: pkg/rbd/nodeserver.go (hunk 1 above)
Comment (Madhu-1) [reply pkg/rbd/nodeserver.go]: it wont override it, `ConstructMountOptions` returns the unique data in opt and req.GetVolumeCapability()

File: e2e/utils.go (hunk 2 above)
Comment (Madhu-1) [reply e2e/utils.go]: made as an argument

**Comment by Madhu-1** [comment]
tested again for  confirmation
```bash
[🎩︎]mrajanna@localhost rbd $]kubectl exec -it csi-rbdplugin-6qzw8 -c csi-rbdplugin sh
sh-4.2# cat /run/
blkid/      console/    faillock/   gluster/    log/        mount/      samba/      sepermit/   systemd/    utmp        
ceph/       cryptsetup/ ganesha/    lock/       lsm/        netreport/  secrets/    setrans/    user/       
sh-4.2# cat /run/mount/utab
utab       utab.lock  
sh-4.2# cat /run/mount/utab
utab       utab.lock  
sh-4.2# cat /run/mount/utab
SRC=/dev/rbd0 TARGET=/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66406bd9-9538-4ff9-bf24-157a2494c5aa/globalmount/0001-0009-rook-ceph-0000000000000001-a723b161-639d-11ea-9bd5-969bf9aa0b35 ROOT=/ OPTS=_netdev
SRC=/dev/rbd0 TARGET=/var/lib/kubelet/pods/1c83e4f6-cd80-40c7-9451-109390c3d531/volumes/kubernetes.io~csi/pvc-66406bd9-9538-4ff9-bf24-157a2494c5aa/mount ROOT=/ BINDSRC=/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66406bd9-9538-4ff9-bf24-157a2494c5aa/globalmount/0001-0009-rook-ceph-0000000000000001-a723b161-639d-11ea-9bd5-969bf9aa0b35 OPTS=_netdev
sh-4.2# 
sh-4.2# 
sh-4.2# 
sh-4.2# mount |grep rbd
tmpfs on /var/lib/kubelet/pods/8fd6c212-3c0c-4010-8546-fa02cd2a1190/volumes/kubernetes.io~secret/rook-csi-rbd-plugin-sa-token-qggq9 type tmpfs (rw,relatime,seclabel)
tmpfs on /var/lib/kubelet/pods/81853df0-25e2-4584-b6a6-2f52fcfeaa35/volumes/kubernetes.io~secret/rook-csi-rbd-provisioner-sa-token-6fxz4 type tmpfs (rw,relatime,seclabel)
tmpfs on /var/lib/kubelet/pods/a7a67406-d877-4fbf-9f50-baf0aad58d5f/volumes/kubernetes.io~secret/rook-csi-rbd-provisioner-sa-token-6fxz4 type tmpfs (rw,relatime,seclabel)
tmpfs on /var/lib/kubelet/pods/cfe1a898-77a0-427a-a59c-85368f93660f/volumes/kubernetes.io~secret/rbd-csi-provisioner-token-jdcxx type tmpfs (rw,relatime,seclabel)
tmpfs on /var/lib/kubelet/pods/ff19954b-3304-4d69-a02e-35662beb18ce/volumes/kubernetes.io~secret/rbd-csi-nodeplugin-token-r455h type tmpfs (rw,relatime,seclabel)
/dev/rbd0 on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-66406bd9-9538-4ff9-bf24-157a2494c5aa/globalmount/0001-0009-rook-ceph-0000000000000001-a723b161-639d-11ea-9bd5-969bf9aa0b35 type ext4 (rw,relatime,seclabel,discard,stripe=1024,data=ordered,_netdev)
/dev/rbd0 on /var/lib/kubelet/pods/1c83e4f6-cd80-40c7-9451-109390c3d531/volumes/kubernetes.io~csi/pvc-66406bd9-9538-4ff9-bf24-157a2494c5aa/mount type ext4 (rw,relatime,seclabel,discard,stripe=1024,data=ordered,_netdev)
sh-4.2# 
```

**Comment by humblec** [comment]
LGTM. Approved. @nixpanic any final comments from you ?

**Comment by humblec** [comment]
@Madhu-1 once this is merged, lets backport it to previous release too.

**Comment by Madhu-1** [comment]
> @Madhu-1 once this is merged, lets backport it to previous release too.

the bot will backport PR once this is merged

**Comment by humblec** [comment]
@nixpanic ping ? :)

**Comment by humblec** [comment]
@ShyamsundarR any review comment here ?

**Comment by Madhu-1** [comment]
@mergifyio refresh

**Comment by Madhu-1** [comment]
@Mergifyio backports release-v2.0

**Comment by Madhu-1** [comment]
@Mergifyio backport release-v2.0

**Pull request merged**